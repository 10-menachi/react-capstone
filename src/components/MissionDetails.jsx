import PropTypes from 'prop-types';
import React from 'react';
import { useDispatch } from 'react-redux';
import { setMissionFree, setMissionReserved } from '../redux/missions/missionsSlice';
import styles from '../styles/Mission.module.css';

const MissionDetails = ({
  name, description, reserved, id,
}) => {
  const dispatch = useDispatch();

  const handleJoin = () => {
    dispatch(setMissionReserved(id));
  };

  const handleLeave = () => {
    dispatch(setMissionFree(id));
  };

  const status = {
    backgroundColor: reserved ? '#379cf6' : '#36454F',
    display: 'block',
    width: '120px',
    border: '1px solid gray',
    color: 'white',
    alignSelf: 'center',
  };

  const row = {
    backgroundColor: reserved ? 'white' : '#E5E4E2',
  };

  return (
    <tr style={row}>
      <td className={styles.name}>{name}</td>
      <td className={styles.description}>{description}</td>
      <td>
        <span style={status}>{reserved ? 'Active Member' : 'Not a Member'}</span>
      </td>
      <td>
        {!reserved && (
          <button type="button" className={styles.join} onClick={handleJoin}>Join Mission</button>
        )}
        {reserved && (
          <button
            type="button"
            className={styles.leave}
            onClick={handleLeave}
          >
            Leave Mission

          </button>
        )}
      </td>
    </tr>
  );
};

MissionDetails.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  reserved: PropTypes.bool,
  id: PropTypes.string.isRequired,
};

MissionDetails.defaultProps = {
  reserved: false,
};

export default MissionDetails;
