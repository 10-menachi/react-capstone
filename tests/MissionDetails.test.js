import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import MissionDetails from '../src/components/MissionDetails';
import missionsReducer from '../src/redux/missions/missionsSlice';

const mockStore = configureStore({
  reducer: {
    missions: missionsReducer,
  },
});
const store = mockStore({});

describe('Tests for MissionDetails component', () => {
  const mission = {
    name: 'Mission name',
    description: 'Mission description',
    id: '1',
    reserved: false,
  };

  it('renders correctly', () => {
    render(
      <Provider store={store}>
        <table>
          <tbody>
            <MissionDetails
              id={mission.id}
              name={mission.name}
              description={mission.description}
              reserved={mission.reserved}
            />
          </tbody>
        </table>
      </Provider>,
    );

    const nameElement = screen.getByText(mission.name);
    const descriptionElement = screen.getByText(mission.description);

    expect(nameElement).toBeInTheDocument();
    expect(descriptionElement).toBeInTheDocument();
  });
});
