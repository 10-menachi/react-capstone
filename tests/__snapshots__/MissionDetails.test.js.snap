// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for MissionDetails component renders correctly 1`] = `
<table>
  <tbody>
    <tr
      style={
        {
          "backgroundColor": "#E5E4E2",
        }
      }
    >
      <td
        className="name"
      >
        Mission name
      </td>
      <td
        className="description"
      >
        Mission description
      </td>
      <td>
        <span
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#36454F",
              "border": "1px solid gray",
              "color": "white",
              "display": "block",
              "width": "120px",
            }
          }
        >
          Not a Member
        </span>
      </td>
      <td>
        <button
          className="join"
          onClick={[Function]}
          type="button"
        >
          Join Mission
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
